lib/nbp/push_connector.rb

```diff
@@ -0,0 +1,115 @@
+# frozen_string_literal: true
+
+module Nbp
+  class PushConnector
+    include Singleton
+
+    class Error < StandardError; end
+    class SettingsError < Error; end
+    class ServerError < Error; end
+
+    def initialize
+      super
+      create_source! unless source_exists?
+    end
+
+    def self.enabled?
+      Settings.nbp&.push_connector&.enable || false
+    end
+
+    def push_lom!(xml)
+      response = api_conn.put("/push-connector/api/lom-v2/#{settings.source.slug}") do |req|
+        req.body = {metadata: xml}.to_json
+      end
+      raise ServerError if response.status == 500
+
+      raise_connector_error('Could not push task LOM', response) unless response.success?
+    end
+
+    def delete_task!(task_uuid)
+      response = api_conn.delete("/push-connector/api/course/#{settings.source.slug}/#{task_uuid}")
+      raise ServerError if response.status == 500
+
+      raise_connector_error('Could not delete task', response) unless response.success? || response.status == 404
+    end
+
+    def source_exists?
+      response = api_conn.get("/datenraum/api/core/sources/slug/#{settings.source.slug}")
+      if response.status == 200
+        true
+      elsif response.status == 404
+        false
+      else
+        raise_connector_error('Could not determine if source exists', response)
+      end
+    end
+
+    def create_source!
+      response = api_conn.post('/datenraum/api/core/sources') do |req|
+        req.body = settings.source.to_json
+      end
+      raise_connector_error('Failed to create source', response) unless response.success?
+    end
+
+    def token
+      if @token.present? && @token_expiration > 10.seconds.from_now
+        @token
+      else
+        update_token
+      end
+    end
```

Comment:
Do we need to consider concurrency here? By default, multiple jobs are executed concurrently. Is each job worker (three by default, if I am right) having an own instance of the `PushConnector` or are they sharing the same? If all workers share the same instance, multiple workers might identify that the token is about to expire and request a new one (resulting in a duplicated token refresh).

Reply:
Yes, the workers all share the same PushConnector since in the default config the workers are spawned as 3 threads in a single process thus sharing global variables. I confirmed this by spawning a 100 jobs and having them print their push connector.
I made the PushConnector thread-safe now with `ThreadLocalVar`. This means each thread maintains an own token but this should not be a problem. What do you think?

Reply:
Thanks, this looks good! While I would have used a mutex, a thread-local variable is a very nice design 👍. 

Besides having one (or two) further questions on concurrency, I am resolving this thread already (there are new dedicated comments).